src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
             Span<byte> t = Span<byte>.Empty;
             Span<byte> remainingOutput = output;
 
-            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+            ReadOnlySpan<byte> infoBuffer = stackalloc byte[0];
+            Span<byte> tempInfoBuffer = stackalloc byte[0];
+            byte[]? rentedTempInfoBuffer = null;
+            const int MaxStackInfoBuffer = 64;
+
+            if (output.Overlaps(info))
             {
-                for (int i = 1; ; i++)
-                {
-                    hmac.AppendData(t);
-                    hmac.AppendData(info);
-                    counter = (byte)i;
-                    hmac.AppendData(counterSpan);
+                tempInfoBuffer = info.Length > MaxStackInfoBuffer ?
+                    (rentedTempInfoBuffer = CryptoPool.Rent(info.Length)).AsSpan(0, info.Length) :
+                    (stackalloc byte[MaxStackInfoBuffer]).Slice(0, info.Length);
```

Comment:
I'm a bit confused about this stackalloc - usually you get stack allocated buffer back when you get out of the current scope which in this case is `}` from the `if` statement which would mean when buffer overlaps you always end up writing where you shouldn't be.

At the same time the variable scope is different so I'm not certain this will work as expected.

cc: @GrabYourPitchforks

Reply:
> usually you get stack allocated buffer back (esp) when you get out of the current scope

Nope, end of the method.  It's why stackalloc in a loop is bad.

```C#
for (int i = 0; i < 1024; i++)
{
    Span<byte> tmp = stackalloc byte[256];
}
```

That's a 256kb stackalloc, split into tiny pieces.

Reply:
Oh, yikes. 🤦 thanks for catching that.

Reply:
Oh, wait - I thought you were saying I was `stackalloc`-ing in a loop, but that does not seem to be the case?

Reply:
I was just giving a counter-example to @krwq 's assertion that stackallocs are popped on block boundaries.  (I guess the easier one would have been that all code like

```C#
Span<byte> something = ...;

if (somethingBoolean)
{
    something = stackalloc byte[20];
}

// use something
```

would be broken)